doc/mid-based-signalling.md

```diff
@@ -0,0 +1,276 @@
+(invite, answer, negotiate).
+
+The `m.call.describe` block is a 'mixin' block in externsible events terms and describes the media being sent on each
+transceiver by the sending user. It *always* refers only to the media being sent by the device that sends the event.
+
+The same is sent either to a focus or a peer client.
+
+A focus will gather the various streams that it's receiving by the `conf_id` from the `m.negotiate` section. In future
+when stream are advertised via the room state, this could be unnecessary and foci need not have any knowledge of what
+group calls are happening (assuming we don't need them to enforce viewship based on this).
+
+The peer or focus then answers, again over to-device message. A peer will start sending media tracks automatically and
+therefore describe them in the answer:
+
+m.call.answer (full mesh)
+```
+{
+    "m.negotiate": {
+        "version": 1,
+        "call_id": "35657a5b793ce",
+        "conf_id": "bbe53499f82e3",
+        "party_id": "678910",
+        "description": {
+            "type": "answer",
+            "sdp": "[...]",
+        }
+    },
+    "m.call.capabilities": {
+        "m.call.transferee": false,
+        "m.call.dtmf": false,
+    },
+    "m.call.describe": {
+        "1": { // transceiver mid 1
+            "media_uuid": "cccc-cccc-cccccc-cccc-cccc",
+            "media_group_uuid": "2345-2345-234567-2345-2345",
+            "purpose": "m.usermedia",
+            "kind": "video"
+        },
+        "2": {
+            "media_uuid": "bbbb-bbbb-bbbbbb-bbbb-bbbb",
+            "media_group_uuid": "1234-1234-123456-1234-1234",
+            "purpose": "m.usermedia",
+            "kind": "audio",
+        },
+    },
+}
+```
+
+A focus, however, will not send any tracks by default and therefore does not include an
+`m.call.describe` bblock. Instead, it includes an `m.track.advertise` block advertising
+what tracks are available for that `conf_id`.
+
+m.call.answer (focus)
+```
+{
+    "m.negotiate": {
+        "version": 1,
+        "call_id": "35657a5b793ce",
+        "conf_id": "bbe53499f82e3",
+        "party_id": "678910",
+        "description": {
+            "type": "answer",
+            "sdp": "[...]",
+        }
+    },
+    "m.call.capabilities": {
+        "m.call.transferee": false,
+        "m.call.dtmf": false,
+    },
+    "m.call.advertise": {
+        "alice:example.org": { // user ID
+            "88888888": { // device ID
+                "2345-2345-234567-2345-2345": [{ // media group uuid
+                    "media_uuid": "aaaa-aaaa-aaaaaa-aaaa-aaaa":
+                    "purpose": "m.usermedia",
+                    "kind": "video",
+                }, {
+                    "media_uuid": "bbbb-bbbb-bbbbbb-bbbb-bbbb":
+                    "purpose": "m.usermedia",
+                    "kind": "audio",
+                },
+            },
+        }
+    },
```

Comment:
From the js-sdk class structure POV, it would make much more sense for this to be structured the same way metadata is atm (`user_id` and `device_id` being props of media group uuid) since the individual `CallFeed`s (media groups) have the `userId`s and `deviceId`s as props, so the mapping makes much more sense that way

Reply:
So basically remove the user ID & device ID levels and move them to attributes of the media group ID? It would be towards the flatter end of possible structures and maybe a nice compromise between depth & duplication. I'm not sure we should necessarily be basing the choices on what the js-sdk does directly, but if the js-sdk has a god reason for doing it that way then that's certainly valid.

Reply:
Yeah, that's what I mean. I'd say the js-sdk doesn't have a **god** reason for this since I was the one who wrote and I haven't gone mad enough to consider myself a diety (yet) :D But it does have a **good** reason - I think it makes sense in terms of media groups being what you display in the UI (but I wrote it so I am biased)

(sorry, I sometimes can't help myself :sweat_smile: )

Reply:
So something like c62296c?